io-engine/src/lvs/lvs_store.rs

```
@@ -415,20 +422,37 @@ impl Lvs {
-                    name: args.name.clone(),
-                    msg: format!("bad metadata resevation ratio: {h}"),
-                })
-            }
+    /// Derive num_md_pages_per_cluster_ratio from max_expansion which can be a factor or absolute size.
+    fn md_resv_ratio(args: &PoolArgs, capacity: u64) -> Result<u32, LvsError> {
+        let param = match args.md_args.as_ref().and_then(|p| p.max_expansion.clone()) {
+            Some(p) => p,
+            None => return Ok(100),
+        };
+        let factor = if let Some(stripped) = param.strip_suffix('x') {
+            stripped
+                .parse::<f64>()
+                .map_err(|error| LvsError::MaxExpansionParse {
+                    msg: format!(
+                        "Failed to parse factor max_expansion {stripped} as float: {error}"
+                    ),
+                })?
+        } else if param.ends_with('B') {
+            let expand_bytes =
+                Byte::from_str(&param).map_err(|error| LvsError::MaxExpansionParse {
+                    msg: format!("Failed to parse max_expansion {param} as bytes: {error}"),
+                })?;
+            expand_bytes.as_u64() as f64 / capacity as f64
         } else {
-            Ok(0)
-        }
+            return Err(LvsError::MaxExpansionParse {
+                msg: format!("Max expansion factor {param} does not end with x or B"),
+            });
+        };
+        // The Blobstore ensures that we have enough pages in used_cluster_mask to track the current device size.
+        // So, If maxExpansion results in < 1x it still shouldn't matter. Its same as having
+        // default reservation. However, it does impact how many md pages per cluster are reserved.
+        // For ex. if the factor turns out to be 0.5 then blobstore reserves 1 page per 2 clusters.
+        // So lets ensures that we pass at least default reservation.
+        Ok((factor.max(1.0) * 100.0) as u32)
```

If we are changing usual default to 200, then this also should be perhaps made 200 here.

---

We eventually decided not have have different default then SPDK. Let user specify themselves if they want to have expansion capability for new Pools. I have mentioned default as 1x in oep as well.

---

> We eventually decided not have have different default then SPDK. Let user specify themselves if they want to have expansion capability for new Pools. I have mentioned default as 1x in oep as well.

So what if someone forgets to specify for some reason but later on intends to expand, and realise he can't?

---

I thought If expansion is critical, users should know and decide what factor they want. Making it default “2x” hides an important configuration knob. They will anyway see `max_expandable_size` and know they have limited expansion capability.